### PR TITLE
API to OkHttpClient to specify an event listener or event listener factory

### PR DESCRIPTION
**Pull request by mjbenedict**
Related to the discussion in #270, this stakes out an API  that could be used to add
Zipkin or other traces to OkHttp requests.

Step one from https://github.com/square/okhttp/pull/3149#issuecomment-277298784

### CODE REVIEW

**Comment by adriancole** [comment]
Thanks for starting. I see only a couple files affected, so will comment on
what I see :)

Seems addX syntax would be used when more than one can be present. Current
impl shows one. Are you planning to provide for multiple event listeners or
is that just a naming glitch?

**Comment by mjbenedict** [comment]
Thanks!  I intend to try to create PRs for what I see as the minimum viable thing and let you guys tell me what needs to be added/changed/fixed.  I am happy to change these to arrays and support multiple event listeners, that seems totally reasonable.

**Comment by adriancole** [comment]
oh.. it wasn't a request to support multiple :) more a naming question
whether that was planned.

**Comment by mjbenedict** [comment]
Yeah I can change this to `setEventListener` and `setEventListenerFactory` 👌

File: okhttp/src/main/java/okhttp3/RealCall.java [hunk 1]
```diff
@@ -42,10 +43,17 @@
   private boolean executed;
 
   RealCall(OkHttpClient client, Request originalRequest, boolean forWebSocket) {
+    final EventListener.Factory eventListenerFactory = client.eventListenerFactory();
+
     this.client = client;
     this.originalRequest = originalRequest;
     this.forWebSocket = forWebSocket;
     this.retryAndFollowUpInterceptor = new RetryAndFollowUpInterceptor(client, forWebSocket);
+    if (null == eventListenerFactory) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/RealCall.java]: nit: we don’t typically do yoda expressions in OkHttp

File: okhttp/src/main/java/okhttp3/OkHttpClient.java [hunk 2]
```diff
@@ -216,6 +216,8 @@ public void apply(ConnectionSpec tlsConfiguration, SSLSocket sslSocket, boolean
   final int readTimeout;
   final int writeTimeout;
   final int pingInterval;
+  final EventListener eventListener;
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/OkHttpClient.java]: I’m unsure about using fields for both of these. Maybe we should have a ConstantFactory or something for the single value case, and when you set the listener that’s what you get?

File: okhttp/src/main/java/okhttp3/OkHttpClient.java [hunk 3]
```diff
@@ -878,6 +896,16 @@ public Builder addNetworkInterceptor(Interceptor interceptor) {
       return this;
     }
 
+    public Builder addEventListener(EventListener listener) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/OkHttpClient.java]: as-is this definitely needs to clear the factory

Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/OkHttpClient.java]: or set the factory to a simple implementation that always returns `listener`

**Comment by swankjesse** [comment]
Awesome. Merged manually!

**Pull request closed without merging**